### PR TITLE
Factory concurrent

### LINKED ISSUE #441: Make factory modules automatically tolerate concurrent application events
At the moment all modules created with a factory module share the same buffer of events that sequentially applies the event. It should be fairly easy to change this at the factory module to have them instead apply events concurrently per module instantiated by the factory.

### PR DESCRIPTION
**Pull request by ranchalp**
Closes #441

### CODE REVIEW

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Thanks for addressing this Alex!
Unfortunately the solution is not as simple as this. The guarantees provided by `ApplyEventsConcurrently` are too weak to be used for parallelizing the factory module.
1. There is no guarantee about the order of application of the events. Thus, events could be applied to submodule in different order than they appear in the input list.
2. Currently, the events destined to submodules and the events for the factory itself (submodule creation and garbage collection) get arbitrarily interleaved, which also might lead to incorrect behavior (e.g. forwarding an event to a garbage-collected submodule).

Also, I would prefer a channel-based implementation over a mutex-based one (I find it much cleaner).

My preferred solution would go something like this:

1. Still use `ApplyEventsSequentially`.
2. When forwarding an event to a submodule, don't apply it directly, but buffer it (add it to an `EventList`, of which there is one per submodule)
3. Whenever an event for the factory itself is processed, and once more at the end, flush the buffers by concurrently calling each submodule's `ApplyEvents` with the respective buffered events, and wait until all those calls return.

This way, the semantics is preserved while exploiting most of the concurrency potential (the overhead of sequentially processing the input and buffering events should be negligible compared to the actual event application). Also note that all submodules that support concurrent event application will do so inside their own `ApplyEvents` methods, but we don't force it on them.

**Comment by ranchalp** [comment]
> 1. There is no guarantee about the order of application of the events. Thus, events could be applied to submodule in different order than they appear in the input list.
> 2. Currently, the events destined to submodules and the events for the factory itself (submodule creation and garbage collection) get arbitrarily interleaved, which also might lead to incorrect behavior (e.g. forwarding an event to a garbage-collected submodule).
> 3. The submodules don't have the option to do concurrent processing themselves, as they are still getting the events one by one.

For clarity: 1 and 2 are the same if I am not mistaken? 

> Whenever an event for the factory itself is processed, and once more at the end, flush the buffers by concurrently calling each submodule's ApplyEvents with the respective buffered events, and wait until all those calls return.

I saw the buffering of events as a separated but desirable issue to be resolved in a different PR, but sure can be done at the same time. I however do not see why sending an event to the factor has anything to do with when the buffers are flushed to the submodules. 

> Also note that all submodules that support concurrent event application will do so inside their own ApplyEvents methods, but we don't force it on them.

The solution I have at the moment does not force concurrency on any submodule. It simply applies sequentialle per submodule using mutexes.

**Comment by ranchalp** [comment]
Done as discussed, feel free to check @matejpavlovic.

**Comment by ranchalp** [comment]
And thanks for the guidelines! :)

File: pkg/factorymodule/factory.go [hunk 1]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
+	// Wait for all goroutines to complete and collect errors and events.
+	for i := 0; i < existingSubmodules; i++ {
+		select {
+		case err := <-errChan:
+			if err != nil {
+				return nil, err
+			}
+		case evtsOut := <-evtsChan:
+			eventsOut.PushBackList(evtsOut)
+		}
```
Comment (matejpavlovic) [inline pkg/factorymodule/factory.go:170]:
```suggestion
		err := <-errChan
		if err != nil {
			return nil, err
		}
		eventsOut.PushBackList(<-evtsChan)
```

File: pkg/util/testlogger/testlogger.go [hunk 2]
```diff
@@ -33,9 +37,12 @@ func (tl *TestLogger) Log(level logging.LogLevel, text string, args ...interface
 		text:  text,
 		args:  args,
 	}
+	tl.internalMutex.Lock()
+	defer tl.internalMutex.Unlock()
 	// TODO: With this implementation, duplicate entries are overwritten. Make duplicates possible.
```
Comment (matejpavlovic) [inline pkg/util/testlogger/testlogger.go:42]: Do we now support duplicate entries? If yes, this comment can be removed.

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Now I like the concept much more, thanks :-)
I added a few implementation-level comments.

File: pkg/factorymodule/factory.go [hunk 3]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
 			return nil, es.Errorf("unsupported event type for factory module: %T", e)
 		}
 	}
-	return fm.forwardEvent(event)
+	fm.bufferSubmoduleEvent(event)
+
+	return events.EmptyList(), nil
+}
+
+// bufferSubmoduleEvent buffers event in a list that is a map where the keys is the moduleID and the value is a list of events
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:121]: Brainfart lol, fixed.

File: pkg/factorymodule/factory.go [hunk 4]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
 }
 
 func (fm *FactoryModule) applyEvent(event *eventpb.Event) (*events.EventList, error) {
-
 	if t.ModuleID(event.DestModule) == fm.ownID {
 		switch e := event.Type.(type) {
 		case *eventpb.Event_Init:
 			return events.EmptyList(), nil // Nothing to do at initialization.
 		case *eventpb.Event_Factory:
+			evtsOut, errs := fm.applySubmodulesEvents()
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:95]: Done.

File: pkg/factorymodule/factory.go [hunk 5]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
 			switch e := factorypbtypes.EventFromPb(e.Factory).Type.(type) {
 			case *factorypbtypes.Event_NewModule:
-				return fm.applyNewModule(e.NewModule)
+				evOut, err := fm.applyNewModule(e.NewModule)
+				if errs != nil {
+					err = errs
+				}
+				return evtsOut.PushBackList(evOut), err
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:102]: Done

File: pkg/factorymodule/factory.go [hunk 6]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
+func (fm *FactoryModule) applySubmodulesEvents() (*events.EventList, error) {
+	// Convenience variable.
+	eventsOut := events.EmptyList()
+	errChan := make(chan error, len(fm.eventBuffer))
+	evtsChan := make(chan *events.EventList, len(fm.eventBuffer))
+
+	existingSubmodules := 0
+	for smID, eventList := range fm.eventBuffer {
+		var submodule modules.PassiveModule
+		var ok bool
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:142]: Nice

File: pkg/factorymodule/factory.go [hunk 7]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
+		if submodule, ok = fm.submodules[smID]; !ok {
+			iter := eventList.Iterator()
+			for event := iter.Next(); event != nil; event = iter.Next() {
+				fm.tryBuffering(event)
+			}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:147]: Done.

File: pkg/factorymodule/factory.go (hunk 1 above)
Comment (ranchalp) [reply pkg/factorymodule/factory.go:170]: Done locally not to have to merge

File: pkg/factorymodule/factory.go [hunk 8]
```diff
@@ -72,29 +74,104 @@ func (fm *FactoryModule) ApplyEvents(evts *events.EventList) (*events.EventList,
...
+		} else {
+			existingSubmodules++
+			go func(submodule modules.PassiveModule, eventList *events.EventList) {
+				evtsOut, err := submodule.ApplyEvents(eventList)
+				if err != nil {
+					errChan <- err
+				} else {
+					evtsChan <- evtsOut
+				}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/factorymodule/factory.go:156]: Done locally not to have to merge

File: pkg/util/testlogger/testlogger.go (hunk 2 above)
Comment (ranchalp) [reply pkg/util/testlogger/testlogger.go:42]: I think we still do not in the testlogger.

**Comment by ranchalp** [comment]
Also just rebased on top of main

**Review by matejpavlovic — APPROVED** [review]
Nice work, all clear for me. 👍

One note on the test logger. This seems to be a textbook example of something to put in another PR as a completely unrelated change. Let's just leave it for now, but for the future, it is always good to ask oneself if all the changes in the code are described by the PR's title and description. If not, they might not belong in that particular PR.

**Comment by matejpavlovic** [comment]
I also added a few lines of comments to the code.
Feel free to merge directly if you agree.

**Comment by ranchalp** [comment]
Thanks! So the problem is that without fixing the test logger the tests would not pass, that is why I decided to put them together in the same PR. Knowing this, do you still think this should go in a separate PR ? Just genuinely asking for next time :).

**Comment by matejpavlovic** [comment]
Oh sorry, I didn't realize that and now it makes much more sense to me. I would still say that the cleanest solution would fixing the logger in a separate PR first, merging it, then rebasing this one on top. But I'm myself guilty of not following this properly in the past. 😬
Let's merge this one and keep improving :-)

**Pull request merged**